DM-38779: Change Butler.ingest() to use resolved DatasetRef

timj opened the pull request:
Previously only unresolved DatasetRef were accepted. Now only resolved DatasetRef can be used.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Change Butler.ingest() to use resolved DatasetRef

Previously only unresolved DatasetRef were accepted.
Now only resolved DatasetRef can be used.

timj committed:
Add repr() for RegistryDefaults

timj commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1862,44 +1862,83 @@ def ingest(
+            dataset_run: str | None = None
             for ref in dataset.refs:
-                if ref.dataId in groupedData[ref.datasetType]:
+                if ref.id is None:
```
Open question: If this is an unresolved ref, should I resolve it using `self.run` or should it be an error?

TallJimbo replied:
Since this question is only relevant during the deprecation period, I think for now we resolve it using the now-deprecated `run` argument, and fall back to `self.run` when that isn't provided.

timj replied:
Okay. I now do resolve the refs -- I was forced into this a bit because the resource gathering graph doesn't use resolved refs and so pipeline execution fails because the `_log` files can't be ingested.

timj committed:
Add ability to resolve unresolved refs in ingest

This is a transitional ability that is needed for old code
using unresolved refs for ingest. This can be triggered by
quantum graphs that use unresolved refs for the log dataset
type.

timj committed:
Modify warning to understand that the datasetType might be supplied externally

timj committed:
Add news fragment

andy-slac commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1807,22 +1809,21 @@ def ingest(
+            relative to the datastore root, if applicable), a resolved
+            `DatasetRef`, and optionally a formatter class or its
+            fully-qualified string name.  If a formatter is not provided, the
+            formatter that would be used for `put` is assumed.  On successful
```
On successful _ingest_?

andy-slac commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1862,44 +1863,97 @@ def ingest(
-        groupedData: GroupedData = defaultdict(dict)
+
+        # We need to reorganize all the inputs so that they are grouped
+        # by dataset type and run. Multiple refs in a single FileDataset
+        # are required to share the run and dataset type.
```
Does `FileDataset` enforce that in general (I don't see changes to `fileDataset.py` on this PR)? Or is this only relevant for ingest?

timj replied:
`FileDataset` does not check that all the refs share a run. Before this they were all unresolved. Didn't cross my mind for the FileDataset constructor to check this (but maybe it should -- future ticket).

timj replied:
Will be resolved in #836.

andy-slac commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1921,35 +1975,65 @@ def ingest(
+
+            # Replace all the refs in the FileDataset with expanded versions.
+            for dataset in grouped_datasets:
+                new_refs = [imported_refs.pop(0) for _ in dataset.refs]
+                dataset.refs = new_refs
```
Made me think more than few instants what `pop` does here. Maybe something more obvious like:

    n_dataset_refs = len(dataset.refs)
    dataset.refs = imported_refs[:n_dataset_refs]
    del imported_refs[:n_dataset_refs]

?

timj replied:
Somehow I merged without seeing this comment.  I can see how your "read the first N" from the list might be clearer than doing N pops... 😄

andy-slac reviewed (approved):
Looks good
